src/nvim/strings.c

```diff
@@ -51,16 +51,14 @@ char_u *vim_strsave(const char_u *string)
   return (char_u *)xstrdup((char *)string);
 }
 
-/*
- * Copy up to "len" bytes of "string" into newly allocated memory and
- * terminate with a NUL.
- * The allocated memory always has size "len + 1", also when "string" is
- * shorter.
- */
+/// Copy up to `len` bytes of `string` into newly allocated memory and
+/// terminate with a NUL. The allocated memory always has size `len + 1`, even
+/// when `string` is shorter.
 char_u *vim_strnsave(const char_u *string, size_t len)
   FUNC_ATTR_NONNULL_RET FUNC_ATTR_MALLOC FUNC_ATTR_NONNULL_ALL
 {
-  return (char_u *)strncpy(xmallocz(len), (char *)string, len);
+  size_t slen = STRLEN(string);
+  return memcpy(xmallocz(len), string, MIN(len, slen));
```

Comment:
This won't work either. It turns out several call sites are depending on the zero-padding behavior of `strncpy`.